Restructure API reference documentation (#10654)
grafana/loki: docs/sources/reference/api.md

---
title: Loki HTTP API
menuTitle: HTTP API
description: Loki exposes HTTP endpoints for data ingestion, data retrieval, as well for cluster management.
aliases:
- ../api/
weight: 100
---

# Loki HTTP API

Loki exposes an HTTP API for pushing, querying, and tailing log data, as well
as for viewing and managing cluster information.

**Note that authorization is not part of the Loki API.**
Authorization needs to be done separately, e.g. using an open-source load-balancer such as NGINX.

> ## Endpoints

Author: I wonder whether we should structure the endpoints overview like in the [Mimir docs](https://grafana.com/docs/mimir/latest/references/http-api/#endpoints).

Reviewer: At first glance I can't tell how it's organized. It doesn't seem to be alphabetized on ANY of the three columns.  Looking closer, it seems to be loosely alphabetized by Endpoint (the third column), which isn't what I would have expected.

I really liked how you chose to structure the endpoints, it seemed very logical to me.  Let's get some feedback from users before we change anything.

### Ingest endpoints

These endpoints are exposed by the `distributor`, `write`, and `all` components:

- [`POST /loki/api/v1/push`](#ingest-logs)

A [list of clients]({{< relref "../send-data" >}}) can be found in the clients documentation.

### Query endpoints

These HTTP endpoints are exposed by the `querier`, `query-frontend`, `read`, and `all` components:

- [`GET /loki/api/v1/query`](#query-logs-at-a-single-point-in-time)
- [`GET /loki/api/v1/query_range`](#query-logs-within-a-range-of-time)
- [`GET /loki/api/v1/labels`](#query-labels)
- [`GET /loki/api/v1/label/<name>/values`](#query-label-values)
- [`GET /loki/api/v1/series`](#query-streams)
- [`GET /loki/api/v1/index/stats`](#query-log-statistics)
- [`GET /loki/api/v1/index/volume`](#query-log-volume)
- [`GET /loki/api/v1/index/volume_range`](#query-log-volume)
- [`GET /loki/api/v1/tail`](#stream-logs)

> API endpoints starting with `/api/prom` are [Prometheus API-compatible](https://prometheus.io/docs/prometheus/latest/querying/api/) and the result formats can be used interchangeably.

Author: I actually removed the `/api/prom/*` endpoints from the list of query endpoints because they are deprecated (we don't remove them from code, only from docs). I didn't see that we still also use `/api/prom/*` endpoints in the ruler API.
@dannykopping Do you think we can also remove them from docs?

> API endpoints starting with `/api/prom` are [Prometheus API-compatible](https://prometheus.io/docs/prometheus/latest/querying/api/) and the result formats can be used interchangeably.

Author: Or should we remove the note about Prometheus compatibility completely? I don't see a point in that anyway.

> API endpoints starting with `/api/prom` are [Prometheus API-compatible](https://prometheus.io/docs/prometheus/latest/querying/api/) and the result formats can be used interchangeably.

Reviewer 2: I think these endpoints are still used extensively by the Grafana Alerting UI and [`cortex-tool`](https://github.com/grafana/cortex-tools/blob/1499f5fb286e672cd238a525169d4cb05f964eab/pkg/commands/rules.go#L245).

### Status endpoints

These HTTP endpoints are exposed by all components and return the status of the component:

- [`GET /ready`](#readiness-probe)
- [`GET /log_level`](#change-log-level)
- [`GET /metrics`](#prometheus-metrics)
- [`GET /config`](#show-current-configuration)
- [`GET /services`](#list-running-services)
- [`GET /loki/api/v1/status/buildinfo`](#show-build-information)

### Ring endpoints

> These HTTP endpoints are exposed by the components that use a consistent hash ring:

Author: `/distributor/ring` is exposed by distributor, and so on ...

> These HTTP endpoints are exposed by the components that use a consistent hash ring:

Author: Suggestion: "These HTTP endpoints are exposed by their respective component that is part of the ring URL prefix."

- [`GET /prometheus/api/v1/rules`](#list-rules)
- [`GET /prometheus/api/v1/alerts`](#list-alerts)

### Log deletion endpoints

These endpoints are exposed by the `compactor`, `backend`, and `all` components:

- [`POST /loki/api/v1/delete`](#request-log-deletion)
- [`GET /loki/api/v1/delete`](#list-log-deletion-requests)
- [`DELETE /loki/api/v1/delete`](#request-cancellation-of-a-delete-request)

### Other endpoints

> These HTTP endpoints are exposed by all components:

Author: Hm, I think I used `all` as an alias for the "single binary" / `all` target. That is a difference to all individual components, though.

- **Terminated**: Service has stopped successfully (terminal state)
- **Failed**: Service has failed in **Starting**, **Running** or **Stopping** state (terminal state)

## Show build information

```
GET /loki/api/v1/status/buildinfo
```

`/loki/api/v1/status/buildinfo` exposes the build information in a JSON object. The fields are `version`, `revision`, `branch`, `buildDate`, `buildUser`, and `goVersion`.

## Flush in-memory chunks to backing store

```
POST /flush
```

`/flush` triggers a flush of all in-memory chunks held by the ingesters to the
backing store. Mainly used for local testing.

In microservices mode, the `/flush` endpoint is exposed by the ingester.

## Prepare ingester shutdown

```
GET, POST, DELETE /ingester/prepare_shutdown
```

This endpoint is used to tell the ingester to release all resources on next SIGTERM.

> After a `POST` to the `prepare_shutdown` endpoint returns, when the ingester process is stopped with `SIGINT` / `SIGTERM`,

Reviewer: After a `POST` to [are we missing a word here?] the `prepare_shutdown` endpoint returns....

> the ingester will be unregistered from the ring and in-memory time series data will be flushed to long-term storage.

Reviewer: ```suggestion
the ingester will be unregistered from the ring, and in-memory time series data will be flushed to long-term storage.
```

```

`/ingester/shutdown` triggers a shutdown of the ingester and notably will _always_ flush any in memory chunks it holds.
This is helpful for scaling down WAL-enabled ingesters where we want to ensure old WAL directories are not orphaned,
but instead flushed to our chunk backend.

It accepts three URL query parameters `flush`, `delete_ring_tokens`, and `terminate`.

**URL query parameters:**

- `flush=<bool>`:
  Flag to control whether to flush any in-memory chunks the ingester holds. Defaults to `true`.
- `delete_ring_tokens=<bool>`:
  Flag to control whether to delete the file that contains the ingester ring tokens of the instance if the `-ingester.token-file-path` is specified. Defaults to `false`.
- `terminate=<bool>`:
  Flag to control whether to terminate the Loki process after service shutdown. Defaults to `true`.

This handler, in contrast to the deprecated `/ingester/flush_shutdown` handler, terminates the Loki process by default.
This behaviour can be changed by setting the `terminate` query parameter to `false`.

In microservices mode, the `/ingester/shutdown` endpoint is exposed by the ingester.

## Distributor ring status

```
GET /distributor/ring
```

> Displays a web page with the distributor hash ring status, including the state, healthy and last heartbeat time of each distributor.

Reviewer: ```suggestion
Displays a web page with the distributor hash ring status, including the state, health, and last heartbeat time of each distributor.
```

## Index gateway ring status

```
GET /indexgateway/ring
```

> Displays a web page with the index gateway hash ring status, including the state, healthy and last heartbeat time of each index gateway.

Reviewer: ```suggestion
Displays a web page with the index gateway hash ring status, including the state, health, and last heartbeat time of each index gateway.
```

  '<compactor_addr>/loki/api/v1/delete?request_id=<request_id>'
```

## Format a LogQL query

```
GET /loki/api/v1/format_query
POST /loki/api/v1/format_query
```

The endpoint accepts the following query parameters in the URL:

- `query`: A LogQL query string. Can be passed as URL param (`?query=<query>`) in case of both `GET` and `POST`. Or as form value in case of `POST`.

> The `/loki/api/v1/format_query` endpoint allows to format LogQL queries. It returns an error if the passed LogQL is invalid. It is exposed by all Loki components and helps to improve readability and the debugging experience of LogQL queries.

Reviewer: ```suggestion
The `/loki/api/v1/format_query` endpoint lets you format LogQL queries. It returns an error if the passed LogQL is invalid. It is exposed by all Loki components and helps to improve readability and the debugging experience of LogQL queries.
```
